rfd/0183-access-request-kube-resource-allow-list.md

```diff
@@ -0,0 +1,204 @@
+
+# User Scenarios
+
+Listing use cases from most permissive to most restrictive:
+
+## As an admin, I don't care what kind of Kubernetes resources a user requests
+
+As a requester, this means I can request for any supported Kubernetes resources (`kube_cluster` and its subresources).
+
+This will be the default behavior if no request options are specified. This will also be the default behavior of existing roles unless modified.
+
+## As an admin, I want to limit what namespaces a user can request
+
+This is already supported. The admin can use `kubernetes_resources` field to restrict what list of namespaces the requester can list and request for:
+
+```yaml
+kind: role
+metadata:
+  name: kube-access-request
+spec:
+  allow:
+    # Only lists namespaces starting with "pumpkin-" prefixes
+    kubernetes_resources:
+      - kind: namespace
+        name: pumpkin-*
+        verbs:
+          - list
+    kubernetes_groups:
+      - '{{internal.kubernetes_groups}}'
+    kubernetes_labels:
+      '*': '*'
+    kubernetes_users:
+      - '{{internal.kubernetes_users}}'
+```
+
+Requesting for namespace not in this list is denied.
+
+## As an admin, I want to limit namespaces using role templates and traits
+
+Trait interpolation will be supported in [v17](https://github.com/gravitational/teleport/pull/45277) for `kubernetes_resources` field:
+
+```yaml
+kind: user
+metadata:
+  name: lisa@goteleport.com
+spec:
+  created_by:
+    connector:
+      id: okta-integration
+  roles:
+  - requester
+  # Attributes from Okta
+  traits:
+    searchable_kube_namespaces:
+    - pumpkin-*
+
+----
+
+kind: role
+metadata:
+  name: kube-access-request
+spec:
+  allow:
+    kubernetes_resources:
+      # Only lists namespaces starting with "pumpkin-" prefixes
+      - kind: namespace
+        # Inserts user's trait value
+        name: '{{external.searchable_kube_namespaces}}'
+        verbs:
+          - list
+    kubernetes_groups:
+      - '{{internal.kubernetes_groups}}'
+    kubernetes_labels:
+      '*': '*'
+    kubernetes_users:
+      - '{{internal.kubernetes_users}}'
+```
+
+## As an admin, I want to require users to request for Kubernetes subresources instead of the whole Kubernetes cluster
+
+Admins can specify in a role, a Kubernetes request mode option that requires users to select [Kubernetes subresources](https://github.com/gravitational/teleport/blob/110b23aefb3c4b91c9d8cca594041b93f0e078dd/api/types/constants.go#L1233).
+
+```yaml
+kind: role
+metadata:
+  name: requester
+spec:
+  allow:
+    request:
+      # Disables requesting whole Kubernetes cluster
+      kubernetes:
+        request_mode: 'resource'
+```
+
+# Details
+
+## New request field for role spec
+
+We will introduce a new role option under the `request` field `kubernetes.request_mode`:
+
+```yaml
+kind: role
+metadata:
+  name: role-name
+spec:
+  allow:
+    request:
+      kubernetes:
+        request_mode: cluster | resource | both
```

Comment:
I would convert it to a list with two modes indicating allowed request modes: `cluster` and `namespace`. Then you would be able to specify `request_modes: ["cluster", "namespace"]` (default) to allow both or individual ones which would be a slightly cleaner interface IMO.

Reply:
i changed it to allow any sub resource kind: https://github.com/gravitational/teleport/blob/ebbdca1c30509f292deadce218d4b8f096a850b4/rfd/0183-access-request-kube-resource-allow-list.md#new-request-field-for-role-spec